tiledb/sm/filesystem/s3.cc

```diff
@@ -1066,6 +1165,171 @@ Status S3::write(const URI& uri, const void* buffer, uint64_t length) {
+  uint64_t nbytes_filled;
+  throw_if_not_ok(fill_file_buffer(buff, buffer, length, &nbytes_filled));
+
+  // Multipart objects will flush whenever the writes exceed file_buffer_size_
+  if (buff->size() == file_buffer_size_) {
+    global_order_write(uri, buff->data(), file_buffer_size_);
+    buff->reset_size();
+  }
+
+  uint64_t bytes_left = length - nbytes_filled;
+  uint64_t offset = nbytes_filled;
+  // Write chunks
+  while (bytes_left > 0) {
+    if (bytes_left >= file_buffer_size_) {
+      global_order_write(uri, (char*)buffer + offset, file_buffer_size_);
+      offset += file_buffer_size_;
+      bytes_left -= file_buffer_size_;
+    } else {
+      throw_if_not_ok(fill_file_buffer(
+          buff, (char*)buffer + offset, bytes_left, &nbytes_filled));
+      offset += nbytes_filled;
+      bytes_left -= nbytes_filled;
+    }
+  }
+}
+
+void S3::global_order_write(
+    const URI& uri, const void* buffer, uint64_t length) {
+  throw_if_not_ok(init_client());
+
+  const Aws::Http::URI aws_uri(uri.c_str());
+  const std::string uri_path(aws_uri.GetPath().c_str());
+
+  UniqueReadLock unique_rl(&multipart_upload_rwlock_);
+  auto state_iter = multipart_upload_states_.find(uri_path);
+  if (state_iter == multipart_upload_states_.end()) {
+    unique_rl.unlock();
+    UniqueWriteLock unique_wl(&multipart_upload_rwlock_);
+
+    state_iter =
+        multipart_upload_states_.emplace(uri_path, MultiPartUploadState())
+            .first;
+
+    unique_wl.unlock();
+
+    bool exists;
+    throw_if_not_ok(is_object(uri, &exists));
+    if (exists) {
+      throw_if_not_ok(remove_object(uri));
+    }
+
+    throw_if_not_ok(initiate_multipart_request(aws_uri, &state_iter->second));
+  } else {
+    // Unlock, as make_upload_part_req will reaquire as necessary.
+    unique_rl.unlock();
+  }
+
+  auto& state = state_iter->second;
+
+  auto& intermediate_chunks = state.buffered_chunks;
+  uint64_t sum_sizes = 0;
+  std::vector<uint64_t> offsets;
+  for (auto& chunk : intermediate_chunks) {
+    offsets.push_back(sum_sizes);
+    sum_sizes += chunk.size;
+  }
+
+  // Account for the current chunk too
+  offsets.push_back(sum_sizes);
+  sum_sizes += length;
+
+  if (sum_sizes < tiledb::sm::constants::s3_min_multipart_part_size) {
+    auto new_uri = generate_chunk_uri(uri, intermediate_chunks.size());
+    throw_if_not_ok(write_direct(new_uri, buffer, length));
+    intermediate_chunks.emplace_back(new_uri.to_string(), length);
+    return;
+  }
+
+  // We have enough buffered for a multipart part upload
+  std::vector<std::byte> merged(sum_sizes);
+  uint64_t length_returned;
+  auto st = parallel_for(
+      vfs_thread_pool_, 0, intermediate_chunks.size(), [&](size_t i) {
+        throw_if_not_ok(read(
+            URI(intermediate_chunks[i].uri),
+            0,
+            merged.data() + offsets[i],
+            intermediate_chunks[i].size,
+            0,
+            &length_returned));
+        return Status::Ok();
+      });
```

Comment:
This took me a while to figure out what we're doing here... Let's add a comment like:
// Read all previously written chunks into one contiguous buffer and add the current data at the end